crates/fs-util/src/lib.rs

```diff
@@ -277,3 +290,59 @@ pub fn write_json_file<T: Serialize>(path: &Path, obj: &T) -> Result<()> {
         .map_err(|source| FsPathError::WriteJson { source, path: path.into() })?;
     writer.flush().map_err(|e| FsPathError::write(e, path))
 }
+
+/// Writes atomically to file.
+///
+/// 1. Creates a temporary file with a `.tmp` extension on the same file directory.
+/// 2. Writes content with `write_fn`.
+/// 3. FSyncs the temp file to disk.
+/// 4. Renames the temp file to the target path.
+/// 5. FSyncs the file directory.
+///
+/// Atomic writes are hard: <https://github.com/paradigmxyz/reth/issues/8622>
+pub fn atomic_write_file<F, E>(file_path: &Path, write_fn: F) -> Result<()>
+where
+    F: FnOnce(&mut File) -> std::result::Result<(), E>,
+    E: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
+    let mut tmp_path = file_path.to_path_buf();
+    tmp_path.set_extension(".tmp");
```

Comment:
i think it will overwrite the original extension, should we rather append it to the end?

Reply:
i thought the expectation is to have `1.wal` turned into `1.wal.tmp`, but it's fine anyway